### PR TITLE
#11786 - Delete immunizations in reduced module when the last other e…

### LINKED ISSUE #11786: Delete immunizations in reduced module when the last other entity related to the person is deleted [2]
### Problem Description

When using the reduced immunization module, as soon as a vaccination is added to a person, that person will never be automatically deleted even if all cases, contacts, event participants and travel entries related to it have already been deleted because the immunization will always stay in the system (since users can't delete them).
 
### Proposed Change

Before permanently deleting persons, for servers using the reduced immunization module, delete all immunizations related to persons that are NOT related to any case, contact, event participant or travel entry.

### Added Value/Benefit

Without this change, permanent person deletion is not working on systems using the reduced immunization system.

### Acceptance Criteria

- [x] A method has been added to ImmunizationService that returns all immunizations related to persons that are not related to any case, contact, event participant or travel entry
- [x] The `deleteOprhanEntities` method in CoreEntityDeletionSerivce has been extended to delete all these immunizations as a batched operation; this needs to be done before deleting the persons
- [x] A unit test covers the new functionality and ensures that on a system with the reduced immunization module active, persons that are only related to immunizations and no other entity are properly deleted

### Implementation Details

Add the immunization deletion to the `CoreEntityDeletionService.deleteOrphanEntities` method.

### Mockups

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by sergiupacurariu**
…ntity related to the person is deleted

Fixes #11786

### CODE REVIEW

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=11954

File: sormas-backend/src/main/java/de/symeda/sormas/backend/immunization/ImmunizationService.java [hunk 1]
```diff
@@ -387,6 +390,47 @@ public Date getLastVaccinationDateBefore(String personUuid, Disease disease, Dat
...
+				cb.not(cb.exists(eventParticipantSubquery)),
+				cb.not(cb.exists(travelEntrySubquery))));
+
+		cq.select(root.get(Immunization.UUID));
+		cq.distinct(true);
+
+		return em.createQuery(cq).getResultList();
+	}
+
+	private Subquery<String> createSubquery(
```
Comment (leventegal-she) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/immunization/ImmunizationService.java:420]: please rename to `createPersonLinkedToOtherCoreEntitySubQuery`

File: sormas-backend/src/main/java/de/symeda/sormas/backend/immunization/ImmunizationService.java [hunk 2]
```diff
@@ -387,6 +390,47 @@ public Date getLastVaccinationDateBefore(String personUuid, Disease disease, Dat
...
+		CriteriaBuilder cb,
+		CriteriaQuery<String> cq,
+		Join<Immunization, Person> personJoin,
+		Class<? extends CoreAdo> subqueryClass,
```
Comment (leventegal-she) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/immunization/ImmunizationService.java:424]: please rename to `otherCoreEntityClass`

File: sormas-backend/src/main/java/de/symeda/sormas/backend/immunization/ImmunizationService.java [hunk 3]
```diff
@@ -387,6 +390,47 @@ public Date getLastVaccinationDateBefore(String personUuid, Disease disease, Dat
...
+		cq.select(root.get(Immunization.UUID));
+		cq.distinct(true);
+
+		return em.createQuery(cq).getResultList();
+	}
+
+	private Subquery<String> createSubquery(
+		CriteriaBuilder cb,
+		CriteriaQuery<String> cq,
+		Join<Immunization, Person> personJoin,
```
Comment (leventegal-she) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/immunization/ImmunizationService.java:423]: please rename to `immunizationPerson`

File: sormas-backend/src/main/java/de/symeda/sormas/backend/immunization/ImmunizationService.java [hunk 4]
```diff
@@ -387,6 +390,47 @@ public Date getLastVaccinationDateBefore(String personUuid, Disease disease, Dat
...
+		Class<? extends CoreAdo> subqueryClass,
+		String personField) {
```
Comment (leventegal-she) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/immunization/ImmunizationService.java:425]: please rename to `otherPersonField`

**Review by leventegal-she — CHANGES_REQUESTED** [review]

**Review by leventegal-she — APPROVED** [review]

**Pull request merged**